### ISSUE TITLE
Infrastructure as code

### ISSUE THREAD

**Issue by tommedema** [AnomalyInnovations/serverless-stack-com#155]
I have written this down more elaborately on [StackOverflow](https://stackoverflow.com/questions/46861678/serverless-framework-ways-to-achieve-full-infrastructure-as-code), but am curious about your view on this.

There is a lot of manual clicking in GUIs (mostly Amazon's console interface) involved when setting up the infrastructure of this app. I.e. the setup is not version controlled and is not easily reproducible. For example the following resources need to be setup manually:

- AWS Incognito User Pool
- AWS Incognite User Pool Application
- AWS Incognito Federated Identity Pool
- AWS DynamoDB instance
- AWS S3 buckets (x3) (this also hosts the [frontend][2])
- AWS CloudFront distribution
- AWS Route53 zone file

How would you take this to the next level where the setup is completely version controlled? The only solutions I can think of would be to either script it, extend the Serverless Framework to spawn not just API Gateways and functions but also the above AWS services, or to wrap everything into a Terraform template (seems complicated).

I couldn't find much on this topic in serverless, surprisingly. Perhaps you could point me in the right direction?

**Comment by tomekrychtyk** [comment]
Maybe AWS CloudFormation would be a good place to start looking for a solution for this? Although it's quite a huge topic...

**Comment by mikepatrick** [comment]
Documenting this would be an excellent addition to the tutorial.  Perhaps a section under "Extra Credit", explaining how[ `resources` and CloudFormation](https://serverless.com/framework/docs/providers/aws/guide/resources/) can help us define even more of our infrastructure as code?

**Comment by jayair** [comment]
Yeah this is definitely on the list. A couple of our readers have taken some swipes at it - https://github.com/AnomalyInnovations/serverless-stack-com/issues/107.

The reason why we haven't done it yet is because CloudFormation does not support the "Email as Username" setting that we use in the Cognito User Pool. It's a new feature and I guess it hasn't been added yet. Though AWS is usually pretty good at being consistent across their platform, so we are hoping that is worked out soon.
